Python interpreter crashes when trying to benchmark zmq client

Hi.
I'm using CentOS7 and am working on a Benchmark project, where one of my tests is to mock a large number of clients that open a zmq.REQ Socket, and start communicating with the server (zmq.REP).

On windows I've seen that the limit of Sockets is 1022 no matter what I do, so I decided to migrate my tests into Linux, and allowing more sockets simply by using `ulimit -n`.

Normally, I would get these on windows:
`FAILED benchmark_static_env.py::TestLivingEnv::test_zmq - zmq.error.ZMQError: Too many open files
`


BUT, when running the same code on linux via pytest, I get these :

2020-07-05 16:20:47,200 [INFO] Socket number - 324
2020-07-05 16:20:47,201 [INFO] Socket number - 325
Fatal Python error: Aborted

```
Thread 0x00007fdb2c1e2700 (most recent call first):
  File "/usr/local/lib/python3.6/site-packages/paramiko/packet.py", line 301 in read_all
  File "/usr/local/li2020-07-05 16:20:47,201 [INFO] Socket number - 326
b/python3.6/site-packages/paramiko/packet.py", line 459 in read_message
  File "/usr/local/lib/python3.6/site-packages/paramiko/transport.py", line 2055 in run
  File "/usr/lib64/python3.6/threading.py", line 916 in _bootstrap_inner
  File "/usr/lib64/python3.6/threading.py", line 884 in _bootstrap

Thread 0x00007fdb4261e740 (most recent call first):
  File "/root/git/BenchmarkCore/zmq_handler.py", line 46 in authenticate
  File "/root/git/BenchmarkCore/zmq_handler.py", line 42 in connect
  File "/root/git/BenchmarkCore/strategies.py", line 140 in open_sockets


```
Also, the number of sockets that are actually getting open on LINUX is always changing from each run, still not sure why.

(When I set ulimit -n to 512 it gives out the same error as on windows, higher than that and python crashed).

Any known solution?
Thanks!

--- a/zmq/backend/cython/utils.pyx
+++ b/zmq/backend/cython/utils.pyx
@@ -56,8 +56,9 @@ def curve_keypair():
     cdef char[64] public_key
     cdef char[64] secret_key
     _check_version((4,0), "curve_keypair")
-    with nogil:
-        rc = zmq_curve_keypair (public_key, secret_key)
+    # see huge comment in libzmq/src/random.cpp
+    # about threadsafety of random initialization
+    rc = zmq_curve_keypair(public_key, secret_key)
     _check_rc(rc)
     return public_key, secret_key
 
@@ -86,8 +87,9 @@ def curve_public(secret_key):
     cdef char[64] public_key
     cdef char* c_secret_key = secret_key
     _check_version((4,2), "curve_public")
-    with nogil:
-        rc = zmq_curve_public (public_key, c_secret_key)
+    # see huge comment in libzmq/src/random.cpp
+    # about threadsafety of random initialization
+    rc = zmq_curve_public(public_key, c_secret_key)
     _check_rc(rc)
     return public_key[:40]
 

